### ISSUE TITLE
Bleadperl breaks File::Slurp 9999.13

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9936]
Migrated from [rt.perl.org#70095](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=70095) (status was 'resolved')

Searchable as RT70095$

**Comment by p5pRT** [comment]
### From @andk

#### Created by sand@k81.nonet


Commit Gitlive\-blead\-1682\-g6100a0a breaks the test suite of File&#8203;::Slurp\.

Author&#8203;: chromatic \<chromatic@&#8203;wgz\.org>
Date&#8203;:   Mon Jul 27 14&#8203;:44&#8203;:15 2009 \+0200

syswrite prints garbage if called with empty scalar and non\-zero offset

:100644 100644 23f79ba4c64a8dc1997a35a244858b02649b9486 ec12cd4aa59e1d39f2626ab83d66ff58864635dc M      pp\_sys\.c

I've reported the issue in https://rt.cpan.org/Ticket/Display.html?id=50939

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl 5.11.0:

Configured by sand at Thu Oct 29 05:48:33 CET 2009.

Summary of my perl5 (revision 5 version 11 subversion 0) configuration:
  Commit id: 6100a0a495d0f5504fc134f242a1ebed97a87087
  Platform:
    osname=linux, osvers=2.6.30-1-amd64, archname=x86_64-linux-ld
    uname='linux k81 2.6.30-1-amd64 #1 smp sat aug 15 18:09:19 utc 2009 x86_64 gnulinux '
    config_args='-Dprefix=/home/src/perl/repoperls/installed-perls/perl/GitLive-blead-1682-g6100a0a -Dinstallusrbinperl=n -Uversiononly -Dusedevel -des -Ui_db -Duselongdouble'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=define
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.3.4', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='long double', nvsize=16, Off_t='off_t', lseeksize=8
    alignbytes=16, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.9.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.9'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    PERL_GIT_UNPUSHED_COMMITS    	/* do not remove this line */
    PERL_GIT_UNCOMMITTED_CHANGES	/* do not remove this line */


@INC for perl 5.11.0:
    /home/src/perl/repoperls/installed-perls/perl/GitLive-blead-1682-g6100a0a/lib/site_perl/5.11.0/x86_64-linux-ld
    /home/src/perl/repoperls/installed-perls/perl/GitLive-blead-1682-g6100a0a/lib/site_perl/5.11.0
    /home/src/perl/repoperls/installed-perls/perl/GitLive-blead-1682-g6100a0a/lib/5.11.0/x86_64-linux-ld
    /home/src/perl/repoperls/installed-perls/perl/GitLive-blead-1682-g6100a0a/lib/5.11.0
    .


Environment for perl 5.11.0:
    HOME=/home/sand
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/sand/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/usr/local/perl/bin:/usr/X11/bin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

-- 
andreas

```
</details>

**Comment by p5pRT** [comment]
### From @gisle

A simpler demonstration of this bug is&#8203;:

$ perl \-le 'syswrite\(STDOUT\, ""\, 0\, 0\)'
Offset outside string at \-e line 1\.

This should be a noop\.

$ perl \-v

This is perl\, v5\.11\.1\-79\-g13f4e2a\* built for darwin\-thread\-multi\-2level

**Comment by p5pRT** [comment]
### From @gisle

I do think that you should be able to use the length of the string as offset without problem 
for strings of any length\.  I think this patch will do the right thing&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/pp_sys.c b/pp_sys.c
index d3430d1..015fcdf 100644
--- a/pp_sys.c
+++ b/pp_sys.c
@@ -1918,7 +1918,7 @@ PP(pp_send)
                    DIE(aTHX_ "Offset outside string");
                }
                offset += blen_chars;
-           } else if (offset >= (IV)blen_chars) {
+           } else if (offset > (IV)blen_chars) {
                Safefree(tmpbuf);
                DIE(aTHX_ "Offset outside string");
            }

Just need to wait for the test suite to finish :-)
```
</details>

**Comment by p5pRT** [comment]
### From @gisle

The related bug that caused that change was http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?
id=67912

**Comment by p5pRT** [comment]
### From @gisle

commit 3c9465285f823f6b117f10189230291c75f08e05
Author&#8203;: Gisle Aas \<gisle@&#8203;aas\.no>
Date&#8203;:   Thu Oct 29 21&#8203;:30&#8203;:02 2009 \+0100

&nbsp;    Allow syswrite with offset same as length of string \[perl \#70095\]
&nbsp;    
&nbsp;    You are not really outside of the string then\, but the effect will
&nbsp;    always be a noop since the number of bytes written will be capped
&nbsp;    to not exceed the length of the string as for other offsets\.

**Comment by p5pRT** [comment]
### From @gisle

I tried to switch the state of this bug to "Resolved" but all I saw was "Ticket 70095&#8203;: Permission 
Denied" :\-\(

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
